src/cxx11.hpp

```diff
@@ -0,0 +1,865 @@
+      }
+
+    template<typename _Clock, typename _Dur, typename _Predicate>
+    bool
+    wait_until(unique_lock<mutex>& __lock, const chrono::time_point<_Clock, _Dur>& __abs_time,
+               _Predicate&& __pred)
+      {
+        while(!(bool) __pred())
+          if(this->wait_until(__lock, __abs_time) == cv_status::timeout)
+            return __pred();
+        return true;
+      }
+
+    template<typename _Rep, typename _Period>
+    cv_status
+    wait_for(unique_lock<mutex>& __lock, const chrono::duration<_Rep, _Period>& __rel_time)
+      {
+        __MCF_ASSERT(__lock.owns_lock());  // must owning a mutex
+        __MCF_ASSERT(__lock.mutex() != nullptr);
+
+        unique_lock<mutex> __temp_lock;
+        __temp_lock.swap(__lock);
+
+        int __err = _Noadl::__wait_for(__rel_time, ::_MCF_cond_wait, this->_M_cnd,
+             ::__MCF_gthr_mutex_unlock_callback, ::__MCF_gthr_mutex_relock_callback,
+             (intptr_t) __temp_lock.mutex()->native_handle(), nullptr);
+
+        __temp_lock.swap(__lock);
+        return (__err == 0) ? cv_status::no_timeout : cv_status::timeout;
+      }
+
+    template<typename _Rep, typename _Period, typename _Predicate>
+    bool
+    wait_for(unique_lock<mutex>& __lock, const chrono::duration<_Rep, _Period>& __rel_time,
+             _Predicate&& __pred)
+      {
+        return this->wait_until(__lock, _Mono_clock::now() + __rel_time, __pred);
+      }
+  };
+
+// Reference implementation for [thread.condition.nonmember]
+inline
+void
+notify_all_at_thread_exit(condition_variable& __cond, unique_lock<mutex> __lock)
+  {
+    __MCF_ASSERT(__lock.owns_lock());  // must owning a mutex
+    __MCF_ASSERT(__lock.mutex() != nullptr);
+
+    __check_thread_atexit(::_MCF_cond_signal_all, __cond.native_handle());
+    __check_thread_atexit(::_MCF_mutex_unlock, __lock.mutex()->native_handle());
+    __lock.release();
+  }
+
+// Reference implementation for [thread.thread.class]
+class thread
+  {
+  private:
+    ::_MCF_thread* _M_thr = nullptr;
+
+    thread(const thread&) = delete;
+    thread& operator=(const thread&) = delete;
+
+  public:
+    constexpr thread() noexcept = default;
+
+    template<typename _Callable, typename... _Args,
+        typename ::std::enable_if<
+            ! ::std::is_same<typename ::std::decay<_Callable>::type, thread>::value
+          >::type* = nullptr>
+    explicit
+    thread(_Callable&& __callable, _Args&&... __args)
+      {
+        enum { __evn_0, __evn_constructed, __evn_cancelled };
+        using _My_invoker = _Invoke_decay_copy<_Callable, _Args...>;
+
+        struct _My_data
+          {
+            ::_MCF_event _M_event[1];
+            _My_invoker _M_invoker[1];
+
+            static
+            void
+            __thunk(::_MCF_thread* __thr)
+              {
+                _My_data* __my = (_My_data*) ::_MCF_thread_get_data(__thr);
+
+                // Check whether `_M_invoker` has been initialized. If its
```

Comment:
Whose constructor? `_M_invoker` is of an array type. I can hardly get the point.

Reply:
This had better be `*_M_invoker`.